In [5]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from glob import glob

#read and put data into data frame
#df = pd.read_csv('station_a.csv')
#print(df)

#print(df.isnull().any(axis=0))

#combine all data
#data_frame = sorted(glob('station_*.csv'))
df_a = pd.read_csv('station_a.csv')
df_b = pd.read_csv('station_b.csv')
df_c = pd.read_csv('station_c.csv')
#new_df = pd.concat([df_a,df_b,df_c],ignore_index='True')
df_new = df_a.merge(df_b,on='time')
df_new2 = df_new.merge(df_c,on='time')
df_main = pd.read_csv('station_main.csv')

#final_df = df_main.merge(new_df, on='time')
final_df = df_new2.merge(df_main,on='time')
final_df.to_csv('final_df.csv')
#concatinate dataframes
#new_df = pd.concat((pd.read_csv(file).assign(filename = file)
          #for file in data_frame), ignore_index = True)
#new_df = pd.me

#new_df.to_csv('new.csv')
final_df

time  temp_c_x status_x  rain_mm_x  temp_c_y   status_y  \
0      2014-01-01 01:00:00       2.6   normal        0.0      -5.7  increased   
1      2014-01-01 02:00:00       2.4   normal        0.0      -2.0     normal   
2      2014-01-01 03:00:00       1.9   normal        0.0      -0.8     normal   
3      2014-01-01 04:00:00       2.0   normal        0.0      -1.6     normal   
4      2014-01-01 05:00:00       1.7   normal        0.0      -2.3  increased   
...                    ...       ...      ...        ...       ...        ...   
34775  2017-12-31 19:00:00       8.9      max        0.1      -0.6        max   
34776  2017-12-31 20:00:00       8.9      max        0.0      -2.3        max   
34777  2017-12-31 21:00:00       8.9      max        0.0      -1.0  increased   
34778  2017-12-31 22:00:00       8.7      max        0.2      -1.7  increased   
34779  2017-12-31 23:00:00       8.4      max        0.6      -1.6  increased   

       rain_mm_y  temp_c     status  rain_mm  level_cm  flow_m2_s  
0            0.0     2.8     normal      0.0     182.0       7.19  
1            0.0     2.5     normal      0.0     182.0       7.19  
2            0.0     1.9     normal      0.0     182.0       7.19  
3            0.0     2.1     normal      0.0     182.0       7.19  
4            0.0     2.1     normal      0.0     182.0       7.19  
...          ...     ...        ...      ...       ...        ...  
34775        0.0     9.0        max      0.0     255.0      15.90  
34776        0.0     9.3  increased      0.0     259.0      16.40  
34777        0.0     9.2        max      0.0     263.0      16.90  
34778        0.0     8.9  increased      0.5     268.0      17.50  
34779        0.0     8.3        max      1.6     272.0      18.00  

[34780 rows x 12 columns]

In [6]:
clean_frame = final_df.dropna()
clean_frame = clean_frame.reset_index(drop=True)
clean_frame

time  temp_c_x status_x  rain_mm_x  temp_c_y   status_y  \
0      2014-01-01 01:00:00       2.6   normal        0.0      -5.7  increased   
1      2014-01-01 02:00:00       2.4   normal        0.0      -2.0     normal   
2      2014-01-01 03:00:00       1.9   normal        0.0      -0.8     normal   
3      2014-01-01 04:00:00       2.0   normal        0.0      -1.6     normal   
4      2014-01-01 05:00:00       1.7   normal        0.0      -2.3  increased   
...                    ...       ...      ...        ...       ...        ...   
31175  2017-12-31 19:00:00       8.9      max        0.1      -0.6        max   
31176  2017-12-31 20:00:00       8.9      max        0.0      -2.3        max   
31177  2017-12-31 21:00:00       8.9      max        0.0      -1.0  increased   
31178  2017-12-31 22:00:00       8.7      max        0.2      -1.7  increased   
31179  2017-12-31 23:00:00       8.4      max        0.6      -1.6  increased   

       rain_mm_y  temp_c     status  rain_mm  level_cm  flow_m2_s  
0            0.0     2.8     normal      0.0     182.0       7.19  
1            0.0     2.5     normal      0.0     182.0       7.19  
2            0.0     1.9     normal      0.0     182.0       7.19  
3            0.0     2.1     normal      0.0     182.0       7.19  
4            0.0     2.1     normal      0.0     182.0       7.19  
...          ...     ...        ...      ...       ...        ...  
31175        0.0     9.0        max      0.0     255.0      15.90  
31176        0.0     9.3  increased      0.0     259.0      16.40  
31177        0.0     9.2        max      0.0     263.0      16.90  
31178        0.0     8.9  increased      0.5     268.0      17.50  
31179        0.0     8.3        max      1.6     272.0      18.00  

[31180 rows x 12 columns]

In [7]:
clean_frame.to_csv('clean_frame.csv')

In [8]:
df_main

time  level_cm  flow_m2_s
0      2014-01-01 01:00:00     182.0       7.19
1      2014-01-01 02:00:00     182.0       7.19
2      2014-01-01 03:00:00     182.0       7.19
3      2014-01-01 04:00:00     182.0       7.19
4      2014-01-01 05:00:00     182.0       7.19
...                    ...       ...        ...
35058  2017-12-31 19:00:00     255.0      15.90
35059  2017-12-31 20:00:00     259.0      16.40
35060  2017-12-31 21:00:00     263.0      16.90
35061  2017-12-31 22:00:00     268.0      17.50
35062  2017-12-31 23:00:00     272.0      18.00

[35063 rows x 3 columns]

Huynh, Thuc Nhat Truong
In the station_main, there is a relation among the data which is the level of water and the flow are realated for example the level_cm is 182 the flow will always be 7.19. The reason could be  there is a stable calculation formular and the flow is dependent on the water level. Therefore we can fill the missing value easily based on the value of level_cm and flow_m2_s. Additionally the uni of flow should be m3/s then it should be changed also. In the code below, I would like to try to firstly replace all the empty value by 0. Afterward, start to fill the missing value.

In [11]:
#replace missing value in station_main:

df_main.to_csv('new_main.csv')

df_main.fillna(0, inplace=True)
df_main.to_csv('new_main2.csv')
df_main = pd.read_csv('new_main2.csv')

# for i in df_main['level_cm']:
#     print(str(count) + " " + str(i))
#     count +=1
for i in range(0,len(df_main)-1):
    if df_main['level_cm'][i] == 0:
        if df_main['flow_m2_s'][i-1] == df_main['flow_m2_s'][i]:
#             df_main['level_cm'][i] = df_main['level_cm'][i-1]
            df_main.at[i,'level_cm'] = df_main['level_cm'][i-1]
        elif df_main['flow_m2_s'][i+1] == df_main['flow_m2_s'][i]:
            df_main.at[i,'level_cm'] = df_main['level_cm'][i+1]
#             df_main['level_cm'][i] = df_main['level_cm'][i+1]   
# print("******")  
for i in range(0,len(df_main)-1):
    if df_main['flow_m2_s'][i] == 0:
        if df_main['level_cm'][i] == df_main['level_cm'][i-1]:
            df_main.at[i,'flow_m2_s'] = df_main['flow_m2_s'][i-1]
#             df_main['flow_m2_s'][i]=df_main['flow_m2_s'][i-1]
        elif df_main['level_cm'][i] == df_main['level_cm'][i+1]:
            df_main.at[i,'flow_m2_s'] = df_main['flow_m2_s'][i+1]
#             df_main['flow_m2_s'][i] = df_main['flow_m2_s'][i+1]   
df_main.to_csv('final_main.csv')